posts/using-swagger-to-build-effective-apis.md

```diff
@@ -0,0 +1,390 @@
+              schema:
+                $ref: '#/components/schemas/Exercise'
+        '400':
+          description: Back-end was unavailable
+```
+The example above shows two GET endpoints that define operations on the exercise resource defined in the components section of the specification.  You'll notice for each endpoint we define an HTTP Method (at least one), any relevant parameters associated with that method request, and the expected responses that come from that request path are mapped to components with the `$ref` tag.
+
+Pathing is a complex topic in API development.  There are many things to consider when defining API paths, like HTTP Headers, HTTP Methods, endpoint names, response types, versions, etc.  For example, you will notice each API path I have specified is a noun.  This noun is either singular or plural, depending on the number of objects returned by the response definition.  The only verbs in this API design are the HTTP verbs, which is as it should be.  Some developers may be tempted to say `http://www.myapi.com/getAllExercises`.  This is not good practice; behind the scenes this operations is a `GET getAllExercises`, which is redundant.  Instead, define your API paths as nouns and use the HTTP operations as verbs.  If you can concatenate the two and create a sensible phrase, you have successfully defined a logical endpoint like `POST exercise` or `GET exercises`.  The use of verbs to define REST API operations is a tricky business and is generally considered bad practice.  However, if you are defining a different kind of API, like SOAP or more generally RPC, this is good practice.  RPC APIs are beyond the scope of this post, but if you would like to do more reading, I would suggest [this article](https://www.smashingmagazine.com/2016/09/understanding-rest-and-rpc-for-http-apis/), which neatly defines the differences between REST APIs and RPC type APIs.
+
+Usually, APIs are written by architects, for developers.  This immediately implies a paradigm shift will be required in order to keep your API future proof from developer requests.  If you want to design your API so it is easily upgradeable, easily maintanable, and easily understood by both users and developers, check out this cool articles on [API Design](https://hackernoon.com/restful-api-design-step-by-step-guide-2f2c9f9fcdbf) and [API Best Practices](https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9).
+
+# Test your Swagger using Swagger Hub tools
+Once you've built your API Specification using the Swagger Editor, you have to test it.  By far, the easiest way you can test your API specification is to copy it from the Swagger Editor and paste it into Swagger Hub.  This will of course require you to sign-in to Swagger Hub and create a new API; but the ease of testing involved make the process well worth it.  Additionally, you can make any new API private on Swagger Hub, if you are concerned others may poach your API.
```

Comment:
```suggestion
Once you've built your API Specification using the Swagger Editor, you have to test it.  By far, the easiest way you can test your API specification is to copy it from the Swagger Editor and paste it into Swagger Hub.  This will, of course, require you to sign-in to Swagger Hub and create a new API; but the ease of testing involved make the process well worth it.  Additionally, you can make any new API private on Swagger Hub, if you are concerned others may poach your API.
```